File: mybox/package/yum_repo.py
import hashlib
from pathlib import Path

from pydantic import Field

from ..configparser import ConfigParser
from ..tracker import Tracker
from ..utils import Optional, raise_
from .manual_version import ManualVersion


class YumRepo(ManualVersion):
    repo_name: str = Field(alias="yum_name")
    baseurl: str = Field(alias="yum_url")
    gpg_key: Optional[str] = None

    def derive_name(self) -> str:
        return f"yum-{self.repo_name}"

    async def get_remote_version(self) -> str:
        m = hashlib.sha256()
        m.update(self.baseurl.encode())
        m.update((self.gpg_key or "").encode())
        return m.hexdigest()

    async def install(self, *, tracker: Tracker) -> None:
        (await self.driver.os()).switch_(
            linux=lambda linux: (
                None
                if linux.distribution == "fedora"
                else raise_(ValueError("YumRepo is only supported on Fedora"))
            ),
            macos=lambda: raise_(ValueError("YumRepo is only supported on Linux")),
        )

        repo = ConfigParser()
        repo[self.repo_name] = {
            "name": self.repo_name,
            "baseurl": self.baseurl,
            "enabled": "1",
        }
        if self.gpg_key is not None:
            repo[self.repo_name]["gpgcheck"] = "1"
            repo[self.repo_name]["gpgkey"] = self.gpg_key

        driver = self.driver.with_root(True)
        path = Path(f"/etc/yum.repos.d/{self.repo_name}.repo")
        await driver.write_file(path, repo.to_string())
        await driver.run("chmod", "a+r", path)
        tracker.track(path, root=True)

        if self.gpg_key:
            await self.driver.with_root(True).run("rpm", "--import", self.gpg_key)

        await super().install(tracker=tracker)

File: src/Mybox/Package/YumRepo.hs
module Mybox.Package.YumRepo (YumRepo (..), mkYumRepo) where

import Data.Text qualified as Text

import Mybox.Aeson
import Mybox.Driver
import Mybox.Package.Class
import Mybox.Package.ManualVersion
import Mybox.Package.Post
import Mybox.Prelude
import Mybox.Tracker

data YumRepo = YumRepo
  { name_ :: Text
  , url :: Text
  , gpgKey :: Maybe Text
  , post :: [Text]
  }
  deriving (Eq, Show)

mkYumRepo :: Text -> Text -> YumRepo
mkYumRepo name url = YumRepo{name_ = name, url, gpgKey = Nothing, post = []}

instance HasField "name" YumRepo Text where
  getField p = "yum-" <> p.name_

instance FromJSON YumRepo where
  parseJSON = withObjectTotal "YumRepo" $ do
    name_ <- takeField "yum_name"
    url <- takeField "yum_url"
    gpgKey <- takeFieldMaybe "gpg_key"
    post <- takePost
    pure YumRepo{..}

instance ToJSON YumRepo where
  toJSON p =
    object $
      [ "yum_name" .= p.name_
      , "yum_url" .= p.url
      , "gpg_key" .= p.gpgKey
      ]
        <> postToJSON p

yumRepoRemoteVersion :: YumRepo -> Text
yumRepoRemoteVersion = jsonEncode

writeRepoFile :: (Driver :> es, Tracker :> es) => YumRepo -> Eff es ()
writeRepoFile p = do
  let repoConfig =
        Text.unlines $
          [ "[" <> p.name_ <> "]"
          , "name=" <> p.name_
          , "baseurl=" <> p.url
          , "enabled=1"
          ]
            ++ ( toList p.gpgKey
                   >>= \key ->
                     [ "gpgcheck=1"
                     , "gpgkey=" <> key
                     ]
               )

  let repoPath = pRoot </> "etc" </> "yum.repos.d" </> (p.name_ <> ".repo")
  drvRun $ sudo $ shellRaw $ "echo " <> shellQuote repoConfig <> " > " <> repoPath.text
  drvRun $ sudo $ "chmod" :| ["a+r", repoPath.text]
  trkAdd p repoPath

importGpgKey :: Driver :> es => YumRepo -> Eff es ()
importGpgKey p = for_ p.gpgKey $ \key ->
  drvRun $ sudo $ "rpm" :| ["--import", key]

yumRepoInstall :: (Driver :> es, Tracker :> es) => YumRepo -> Eff es ()
yumRepoInstall p = do
  os <- drvOS
  case os of
    Linux Fedora -> do
      writeRepoFile p
      importGpgKey p
    Linux (Debian _) -> terror "YumRepo is only supported on Fedora"
    MacOS -> terror "YumRepo is only supported on Linux"

instance Package YumRepo where
  remoteVersion = pure . yumRepoRemoteVersion
  localVersion = manualVersion
  install = installWithPost $ manualVersionInstall yumRepoInstall
